Adding CI for Pull Requests

jeffctown opened the pull request:
# Motivation

I noticed that an open pull request didn't compile, so I thought it would be good to know if any pull requests fail to compile before they are merged into master.  

The goal of this pull request is to help prevent merging in something that will break the master branch.

## What's in This Pull Request?

This pull request adds a simple [Github Action](https://github.com/features/actions) to build the app on any pull request.

I'm not sure if any CI exists for this repo yet, so let me know if there is a preference for another CI system.  [Github Actions is free for open source repositories](https://github.com/features/actions).

## When Does it Run?

Once this PR is merged to master, this will run an xcodebuild on any new pull request or push to an existing pull request that is targeted to the master branch.

## What Happens Then?

1. Once you put up a PR, you'll notice it is running a status check:

<img width="753" alt="Screen Shot 2020-04-08 at 8 02 59 AM" src="https://user-images.githubusercontent.com/2142301/78783327-75a16c00-7971-11ea-94b1-c003ff7b4801.png">

**NOTE: Notice that right now the merge button is enabled during the status check.  We probably don't want that.**

2. If you click on Details, you will be taken to a Github Actions page to monitor the progress of the build:

<img width="1673" alt="Screen Shot 2020-04-08 at 8 04 20 AM" src="https://user-images.githubusercontent.com/2142301/78783506-bbf6cb00-7971-11ea-98e6-c2bed86d0dc0.png">

3. Once the build passes, you'll see the UI updated for this PR:

<img width="1680" alt="Screen Shot 2020-04-08 at 8 10 35 AM" src="https://user-images.githubusercontent.com/2142301/78783883-548d4b00-7972-11ea-9e73-eabe79797607.png">
<img width="298" alt="Screen Shot 2020-04-08 at 8 10 45 AM" src="https://user-images.githubusercontent.com/2142301/78783678-0710de00-7972-11ea-983a-99ce6e811c65.png">

<img width="771" alt="Screen Shot 2020-04-08 at 8 10 53 AM" src="https://user-images.githubusercontent.com/2142301/78783676-06784780-7972-11ea-9814-94e6c807badd.png">

## Should this prevent merging?

Yes, normally you want to prevent merging in a branch that won't compile.  This setting should be enabled in the repository's settings.  I believe you would need to enable the "require status checks to pass" if you wanted to prevent merging before a pull request has a successful build:

<img width="1031" alt="Screen Shot 2020-04-08 at 8 05 14 AM" src="https://user-images.githubusercontent.com/2142301/78783246-4d197200-7971-11ea-9697-5d51de4c55ea.png">

If this is enabled, nobody will be able to merge a pull request until a successful build has completed. 👍

(I don't have the permission in this repo to add any reviewers)

jeffctown commented on .github/workflows/pull_request.yml:
```diff
@@ -0,0 +1,17 @@
+  pull_request:
+    branches: [ master ]
+env:
+  DEVELOPER_DIR: /Applications/Xcode_11.4.app/Contents/Developer
```
Set to use Xcode 11.4

ibeckermayer replied:
How does this DEVELOPER_DIR work? Is that referencing some virtual machine being run by `Github Actions`?

jeffctown replied:
This is not very well documented at all, so I had to do some digging too.  Sorry for not explaining this already!

I found these posts: 
https://www.jessesquires.com/blog/selecting-an-xcode-version-on-github-ci/
https://github.community/t5/GitHub-Actions/Selecting-an-Xcode-version/td-p/31105

I would imagine Github will add a more explicit way to set Xcode version, but this is how it's apparently supposed to be done today.

jeffctown replied:
Updated [here](https://github.com/covid19risk/covidwatch-ios/pull/20/commits/b11773639a5e75a6fb5381ce5e3cb0f94ccb0649).

jeffctown commented on .github/workflows/pull_request.yml:
```diff
@@ -0,0 +1,17 @@
+
+on:
+  pull_request:
+    branches: [ master ]
```
Building PRs targeting master only

ibeckermayer replied:
For now we are developing off of `dev` so we should add that as well.

jeffctown replied:
Good catch! I'll update this.

jeffctown replied:
Updated [here](https://github.com/covid19risk/covidwatch-ios/pull/20/commits/b11773639a5e75a6fb5381ce5e3cb0f94ccb0649).

ibeckermayer commented:
Hey Jeff this looks great. Thank you for providing such a detailed explanation of GitHub Actions. I've used CircleCI for something similar but never touched Actions, looks like a solid tool though. I left a couple of comments above and am going to talk with Rhys to get you the permissions necessary to toggle the settings.

Also I just saw your message in Slack (apologies, this Slack can be a bit hectic to keep up with at the moment). We definitely should throw a `swiftlint` or `swiftformat` into the mix. I'm new to iOS dev and so am un-opinionated on that; feel free to make the call. Also feel free to throw down some standards (i.e. no force unwrapping etc.) and we can discuss the merits/drawbacks of those in a meeting with the other iOS devs.

As for Storyboard, I think that ship has sailed for now. I learned Storyboard (and Swift in general) for this project and was using that originally, but once I was got into anything but trivial layouts it started causing some major headaches. For now I've been trying to write the UI in the "component" style I'm used to (Vue), and have been using proportions from Figma to deal with scaling. We've got it working down to the iPhone 8 but haven't tested it on anything smaller than that due to the lack of simulators in XCode. I'm still convinced that I'll be able to make it work at different sizes relatively elegantly this way, but willing to discuss further if you think I'm making a major mistake.

jeffctown commented:
@ibeckermayer - I think I've addressed all of your comments, but let me know if I missed anything.  This looks ready to 🚢 to me. 

I'll look into SwiftLint next.  

II do have a strong opinion that we should be using auto layout for this project, but I think it's easier to discuss in Slack.

jeffctown commented:
Handled in #23.
